FreeMonoid.agda

```diff
@@ -0,0 +1,72 @@
+module FreeMonoid where
+
+open import Monoid
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality
+import Relation.Binary.SetoidReasoning as ≈-Reasoning
+
+
+module Carrier (A : Set) where
+
+  data ExpMonoid : Set where
+    Embed : A → ExpMonoid
+    MEmpty : ExpMonoid
+    _<>_ : ExpMonoid → ExpMonoid → ExpMonoid
+
+
+  data EquivMonoid : ExpMonoid → ExpMonoid → Set where
+    equiv-refl : {e : ExpMonoid}
+               → EquivMonoid e e
+    equiv-sym : {e₁ e₂ : ExpMonoid}
+              → EquivMonoid e₁ e₂ → EquivMonoid e₂ e₁
+    equiv-trans : {e₁ e₂ e₃ : ExpMonoid}
+                → EquivMonoid e₁ e₂
+                → EquivMonoid e₂ e₃
+                → EquivMonoid e₁ e₃
+    equiv-cong : {e₁ e₂ e₁′ e₂′ : ExpMonoid}
+               → EquivMonoid e₁ e₁′
+               → EquivMonoid e₂ e₂′
+               → EquivMonoid (e₁ <> e₂) (e₁′ <> e₂′)
+    left-id-mempty : {e : ExpMonoid}
+                   → EquivMonoid (MEmpty <> e) e
+    right-id-mempty : {e : ExpMonoid}
+                    → EquivMonoid (e <> MEmpty) e
+    assoc-mappend : (e₁ e₂ e₃ : ExpMonoid)
+                  → EquivMonoid ((e₁ <> e₂) <> e₃) (e₁ <> (e₂ <> e₃))
+
+  equiv-setoid : Setoid _ _
+  equiv-setoid = record
+    { Carrier = ExpMonoid
+    ; _≈_ = EquivMonoid
+    ; isEquivalence = record
+      { refl  = equiv-refl
+      ; sym   = equiv-sym
+      ; trans = equiv-trans
+      }
+    }
+
+
+  eqClass : ExpMonoid → List A
+  eqClass = {!!}
+
+
+  module Sound where
```

Comment:
Are you familiar with `≈-Reasoning`? It's the first time I use it, I found it very convenient in terms of proof readability. As you can see in the example below, it's very similar to `≡-Reasoning`, except that in addition to `_≡⟨_⟩_` you can also use `_≈⟨_⟩_` to provide a proof of equivalence, which in this case would be a value of type `EquivMonoid e₁ e₂`.